flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/nn/KNN.scala

```diff
@@ -0,0 +1,321 @@
+                    training.values.head.size + math.log(math.log(training.values.length) /
+                      math.log(4.0)) < math.log(training.values.length) / math.log(4.0) &&
+                      (metric.isInstanceOf[EuclideanDistanceMetric] ||
+                        metric.isInstanceOf[SquaredEuclideanDistanceMetric]))
+
+                  if (useQuadTree) {
+                    knnQueryWithQuadTree(training.values.asInstanceOf[Vector[DenseVector]],
+                      testing.values,k, metric,queue, out)
+                  } else {
+                    knnQueryBasic(training.values.asInstanceOf[Vector[DenseVector]],
+                      testing.values,k, metric,queue, out)
+                  }
+                }
+              }
+            }
+
+            // group by input vector id and pick k nearest neighbor for each group
+            val result = crossed.groupBy(2).sortGroup(3, Order.ASCENDING).reduceGroup {
+              (iter, out: Collector[(FlinkVector, Array[FlinkVector])]) => {
+                if (iter.hasNext) {
+                  val head = iter.next()
+                  val key = head._2
+                  val neighbors: ArrayBuffer[FlinkVector] = ArrayBuffer(head._1)
+
+                  for ((vector, _, _, _) <- iter.take(k - 1)) {
+                    // we already took a first element
+                    neighbors += vector
+                  }
+
+                  out.collect(key, neighbors.toArray)
+                }
+              }
+            }
+
+            result
+          case None => throw new RuntimeException("The KNN model has not been trained." +
+            "Call first fit before calling the predict operation.")
+
+        }
+      }
+    }
+  }
+
+  def knnQueryWithQuadTree[T <: FlinkVector](training: Vector[DenseVector],
+                           testing: Vector[(Long, T)],
+                           k: Int, metric: DistanceMetric,
+                           queue: mutable.PriorityQueue[(FlinkVector, FlinkVector, Long, Double)],
+                           out: Collector[(FlinkVector, FlinkVector, Long, Double)]) {
+    /// find a bounding box
+    val MinArr = List.range(0, training.head.size).toArray
+    val MaxArr = List.range(0, training.head.size).toArray
+
+    val minVecTrain = MinArr.map(i => training.map(x => x(i)).min - 0.01)
+    val minVecTest = MinArr.map(i => testing.map(x => x._2(i)).min - 0.01)
+    val maxVecTrain = MaxArr.map(i => training.map(x => x(i)).min + 0.01)
+    val maxVecTest = MaxArr.map(i => testing.map(x => x._2(i)).min + 0.01)
+
+    val MinVec = DenseVector(MinArr.map(i => Array(minVecTrain(i), minVecTest(i)).min))
+    val MaxVec = DenseVector(MinArr.map(i => Array(maxVecTrain(i), maxVecTest(i)).max))
+
+    var trainingQuadTree = new QuadTree(MinVec, MaxVec, metric)
+
+    if (trainingQuadTree.maxPerBox < k) { /// make sure at least k points/box
+      trainingQuadTree.maxPerBox = k
```

Comment:
Done -- now in the constructor and now pass in `val maxPerBox = Array(k,20).max` into the constructor in KNN.scala